### PR TITLE
Add DescriptorLoadOp and DescriptorStoreOp lowering in `LoadStoreOpToLLVM` pass

### PR DESCRIPTION
**Pull request by anthonycanino**
Add load/store for DescriptorLoadOp and DescriptorStoreOp.

Lowering is for gather load/store only, and is analgous to the
existing LoadOp/StoreOp lowering, but adjusted for TensorDescriptor.

Some common pathways have been refactored, notably, the logic of
`convertBlockPtrToTensorOfPtr` to work with both block ptr input
and tensor descriptor input.

There are a couple spots that will need to be amended later on:

1. Right now, padding information is attached to the `MakeTensorDescOp` and does not propagate to the type or through function calls. We need to flow this information to the `DescriptorLoadOp` through function invocations (not sure if its worth an analysis pass or if there is a simpler way to accomplish this).

2. We need axis analysis for TensorDescriptor ops to perform the vectorization.

### CODE REVIEW

**Comment by anthonycanino** [comment]
@etiotto @whitneywhtsang I think I am hitting some issues with the existing `RewriteTensorDescriptorToPointer.cpp` pass.

So it makes sense to me that we are decomposing a `TensorDescriptor` to its components and passing to the generated kernel which expects the same (when we have `RewriteTensorDescriptorToPointer` active; however, this pass obviously removes the TensorDescriptors which means I cannot natively handle them in `LoadStoreOpToLLVM.cpp`. 

How would we approach this PR? Since many components have to be implemented in order to test on the full integration suite. Would we instead add some unit tests to test my lowerings (which do run locally, and will run from python with a .ttgir I have written).

**Comment by whitneywhtsang** [comment]
There are a lot of passes need to be enhanced to support tensor descriptor natively end to end. Please test your lowering with lit test and/or tests like `python/test/unit/intel/test_block_io.py` which has TTGIR as input.

**Comment by whitneywhtsang** [comment]
Note that after https://github.com/intel/intel-xpu-backend-for-triton/pull/6103, DescriptorLoadOp and DescriptorStoreOp won't be rewritten in RewriteTensorDescriptorToPointer, but they will still be rewritten by TritonIntelTensorDescToBlockPointer. The plan is to move TritonIntelTensorDescToBlockPointer later in the pipeline progressively until we can completely remove it.

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:1]: suggest to have one PR that handle non 2d block io codegen, and another PR that handle 2d block io codegen.

Comment (anthonycanino) [reply third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:1]: Yes, I agree https://github.com/intel/intel-xpu-backend-for-triton/issues/5679#issuecomment-3945271869

File: third_party/intel/lib/TritonIntelGPUToLLVM/TypeConverter.cpp [hunk 1]
```diff
@@ -32,11 +32,29 @@ static Type convertTritonPointerType(triton::PointerType type) {
...
 
+static Type convertTritonDescType(triton::TensorDescType type) {
+  auto ctx = type.getContext();
+  auto blockType = type.getBlockType();
+  // struct { shape0, shape1, ..., stride0, stride1, ..., base_ptr }
+  auto shape = blockType.getShape();
+  SmallVector<Type, 4> types;
+  // shapes, strides
+  for (size_t i = 0; i < 2 * shape.size(); ++i)
+    types.push_back(IntegerType::get(ctx, 64));
```
Comment (chengjunlu) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TypeConverter.cpp:43]: Align the type to the triton operation `TT_MakeTensorDescOp`.

```
let arguments = (ins
    TT_Ptr:$base,
    Variadic<I32>:$shape,
    Variadic<I64>:$strides,
    DefaultValuedAttr<TT_PaddingOptionAttr, "::mlir::triton::PaddingOption::PAD_ZERO">:$padding
  );
```

Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TypeConverter.cpp:43]: The shape has already been extended to i64 in constructing the struct. 
```
for (auto shape : shapes)
      elems.push_back(b.sext(i64_ty, shape));
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/TensorDescOpsToLLVM.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    auto shapes = adaptor.getShape();
+    auto strides = adaptor.getStrides();
+    auto base = adaptor.getBase();
+    auto result = op.getResult();
+
+    SmallVector<Value> elems;
+
+    // Shapes: extend from i32 to i64
+    for (auto shape : shapes)
+      elems.push_back(b.sext(i64_ty, shape));
```
Comment (chengjunlu) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TensorDescOpsToLLVM.cpp:29]: The shape should be un-signed. It is meaningless if the shape is nagetive. 
We can use `zext`

Comment (anthonycanino) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TensorDescOpsToLLVM.cpp:29]: Happy to make the change, but for reference, this is following the pattern seen in https://github.com/intel/intel-xpu-backend-for-triton/blob/main/third_party/intel/lib/TritonIntelGPUToLLVM/TensorPtrOpsToLLVM.cpp#L40-L52

Do we still want to adjust?

File: python/test/unit/intel/test_block_tdesc.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            tt.descriptor_store %dst_desc [%c0_i32, %c0_i32], %data
+                                : !tt.tensordesc<tensor<{M}x{N}x{ty}, #blocked>>, tensor<{M}x{N}x{ty}, #blocked>
+
+            tt.return
+        }}
+    }}
+    """
+
+    torch_dtype = getattr(torch, dtype_str)
+    if torch_dtype.is_floating_point:
```
Comment (etiotto) [inline python/test/unit/intel/test_block_tdesc.py:56]: [SUGGESTION]: `torch.randn()` and `torch.randint()` do not use a seed. Add `torch.manual_seed(42)` before the if.

File: python/test/unit/intel/test_block_tdesc.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        monkeypatch.setenv("TRITON_INTEL_PREDICATED_LOAD", "1")
+        monkeypatch.setenv("TRITON_INTEL_PREDICATED_STORE", "1")
+    yield
+
+
+@pytest.mark.parametrize("M, N",
+                         [[256, 64], [256, 32], [128, 32], [128, 16], [64, 64], [64, 32], [32, 32], [16, 64], [16, 16]])
+@pytest.mark.parametrize("dtype_str", ["float32", "float16", "int8"])
+@pytest.mark.skipif(not is_xpu(), reason="Block descriptor tests are specific to the XPU backend")
+def test_block_tdesc_load_store(M, N, dtype_str, device, tmp_path: pathlib.Path):
```
Comment (whitneywhtsang) [inline python/test/unit/intel/test_block_tdesc.py]:
```suggestion
@pytest.mark.skipif(not is_xpu(), reason="Tensor descriptor tests are specific to the XPU backend")
def test_tdesc_load_store(M, N, dtype_str, device, tmp_path: pathlib.Path):
```

File: python/test/unit/intel/test_block_tdesc.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+    torch_dtype = getattr(torch, dtype_str)
+    if torch_dtype.is_floating_point:
+        a = torch.randn((M, N), dtype=torch_dtype, device=device)
+    else:
+        a = torch.randint(low=-127, high=128, size=(M, N), dtype=torch_dtype, device=device)
+
+    x = torch.empty_like(a)
+
+    temp_file = tmp_path / "test_block_tdesc_load_store.ttgir"
```
Comment (whitneywhtsang) [inline python/test/unit/intel/test_block_tdesc.py]:
```suggestion
    temp_file = tmp_path / "test_tdesc_load_store.ttgir"
```

File: python/test/unit/intel/test_block_tdesc.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+    kernel[(1, 1, 1)](a, x)
+    assert torch.equal(a, x)
+
+
+@pytest.mark.parametrize("M, N",
+                         [[256, 64], [256, 32], [128, 32], [128, 16], [64, 64], [64, 32], [32, 32], [16, 64], [16, 16]])
+@pytest.mark.parametrize("dtype_str", ["float32", "float16", "int8"])
+@pytest.mark.skipif(not is_xpu(), reason="Block descriptor tests are specific to the XPU backend")
+def test_block_tdesc_load_zero_padding(M, N, dtype_str, device, tmp_path: pathlib.Path):
```
Comment (whitneywhtsang) [inline python/test/unit/intel/test_block_tdesc.py]:
```suggestion
@pytest.mark.skipif(not is_xpu(), reason="Tensor descriptor tests are specific to the XPU backend")
def test_tdesc_load_zero_padding(M, N, dtype_str, device, tmp_path: pathlib.Path):
```

File: python/test/unit/intel/test_block_tdesc.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            tt.return
+        }}
+    }}
+    """
+
+    torch_dtype = getattr(torch, dtype_str)
+    a = torch.ones((M, N), dtype=torch_dtype, device=device)
+    x = torch.empty_like(a)
+
+    temp_file = tmp_path / "test_block_tdesc_load_zero_padding.ttgir"
```
Comment (whitneywhtsang) [inline python/test/unit/intel/test_block_tdesc.py]:
```suggestion
    temp_file = tmp_path / "test_tdesc_load_zero_padding.ttgir"
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 8]
```diff
@@ -2308,6 +2353,349 @@ struct LoadOpConversion : public ConvertOpToLLVMPattern<triton::LoadOp>,
...
+    // Try to get the padding option from the defining MakeTensorDescOp.
+    // NOTE: This method only works when the descriptor is defined locally
+    // (i.e., not passed through block arguments, function arguments, or
+    // control flow). For descriptors that flow through control flow, we would
+    // need an analysis pass to propagate tensor descriptor information.
+    // TODO: Implement an analysis pass to propagate MakeTensorDescOp info
+    // through control flow for non-local descriptor definitions.
+    std::optional<PaddingOption> padding = std::nullopt;
+    if (auto makeDescOp =
+            op.getDesc().getDefiningOp<triton::MakeTensorDescOp>()) {
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp]: Should we use findDefiningOpOfType in third_party/intel/include/Utils/Utility.h instead?

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 9]
```diff
@@ -2308,6 +2353,349 @@ struct LoadOpConversion : public ConvertOpToLLVMPattern<triton::LoadOp>,
...
+      padding = makeDescOp.getPadding();
+    }
+    // If padding is not found (descriptor came through control flow),
+    // default to PAD_ZERO for safety.
+    if (!padding)
+      padding = PaddingOption::PAD_ZERO;
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp]: how about initializing padding with PaddingOption::PAD_ZERO?

File: third_party/intel/lib/TritonIntelGPUToLLVM/TensorDescOpsToLLVM.cpp (hunk 2 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TensorDescOpsToLLVM.cpp:29]:
```
    for (auto shape : shapes)
      elems.push_back(shape);
```

Do you want to refer a different place at first?

**Review by chengjunlu — APPROVED** [review]

**Review by whitneywhtsang — APPROVED** [review]

**Comment by anthonycanino** [comment]
I think https://github.com/intel/intel-xpu-backend-for-triton/pull/5759 will affect this PR now (in the sense, I should probably mirror that functionality).

**Comment by anthonycanino** [comment]
> Blocking accidental merge due to [#6118 (comment)](https://github.com/intel/intel-xpu-backend-for-triton/pull/6118#issuecomment-3967569681)

@whitneywhtsang I was mistaken --- it is the PR I am currently working on (block io) that would be affected, not this one.

**Comment by anthonycanino** [comment]
@whitneywhtsang can this be merged?
